contracts/verifying/VerifyingSingletonPaymasterV2.sol

```diff
@@ -0,0 +1,390 @@
+        //don't revert on signature failure: return SIG_VALIDATION_FAILED
+        if (
+            verifyingSigner != hash.toEthSignedMessageHash().recover(signature)
+        ) {
+            // empty context and sigFailed with time range provided
+            return (context, _packValidationData(true, validUntil, validAfter));
+        }
+
+        require(priceMarkup <= 2e6, "Verifying PM:high markup %");
+
+        uint32 dynamicMarkup = maxuint32(priceMarkup, fixedPriceMarkup);
+        require(dynamicMarkup >= 1e6, "Verifying PM:low markup %");
+
+        uint256 effectiveCost = (requiredPreFund * dynamicMarkup) /
+            PRICE_DENOMINATOR;
+
+        if (effectiveCost > paymasterIdBalances[paymasterId])
+            revert InsufficientBalance(
+                effectiveCost,
+                paymasterIdBalances[paymasterId]
+            );
+
+        context = abi.encode(
+            paymasterId,
+            priceMarkup,
+            userOp.maxFeePerGas,
+            userOp.maxPriorityFeePerGas
+        );
+
+        return (context, _packValidationData(false, validUntil, validAfter));
+    }
+
+    function parsePaymasterAndData(
+        bytes calldata paymasterAndData
+    )
+        public
+        pure
+        returns (
+            address paymasterId,
+            uint48 validUntil,
+            uint48 validAfter,
+            uint32 priceMarkup,
+            bytes calldata signature
+        )
+    {
+        (paymasterId, validUntil, validAfter, priceMarkup) = abi.decode(
+            paymasterAndData[VALID_PND_OFFSET:SIGNATURE_OFFSET],
+            (address, uint48, uint48, uint32)
+        );
+
+        signature = paymasterAndData[SIGNATURE_OFFSET:];
+    }
+
+    /**
+     * @dev Executes the paymaster's payment conditions
+     * @param mode tells whether the op succeeded, reverted, or if the op succeeded but cause the postOp to revert
+     * @param context payment conditions signed by the paymaster in `validatePaymasterUserOp`
+     * @param actualGasCost amount to be paid to the entry point in wei
+     */
+    function _postOp(
+        PostOpMode mode,
+        bytes calldata context,
+        uint256 actualGasCost
+    ) internal virtual override {
+        (
+            address paymasterId,
+            uint32 priceMarkup,
+            uint256 maxFeePerGas,
+            uint256 maxPriorityFeePerGas
+        ) = abi.decode(context, (address, uint32, uint256, uint256));
+
+        uint256 effectiveGasPrice = getGasPrice(
+            maxFeePerGas,
+            maxPriorityFeePerGas
+        );
+
+        uint32 dynamicMarkup = maxuint32(priceMarkup, fixedPriceMarkup);
+
+        uint256 balToDeduct = actualGasCost +
+            unaccountedEPGasOverhead *
+            effectiveGasPrice;
+
+        uint256 costIncludingPremium = (balToDeduct * dynamicMarkup) /
+            PRICE_DENOMINATOR;
+
+        // Cache storage reads
+        uint256 paymasterBalance = paymasterIdBalances[paymasterId];
```

Comment:
<img width="776" alt="Screenshot 2023-11-06 at 3 47 21 PM" src="https://github.com/bcnmy/biconomy-paymasters/assets/16562513/faaa1c38-66d9-435c-85b2-3677f114be07">

>gas used: 27646

<img width="739" alt="Screenshot 2023-11-06 at 3 49 01 PM" src="https://github.com/bcnmy/biconomy-paymasters/assets/16562513/0d7c9686-4983-44b3-80aa-8a72256062e8">

>gas used: 27767

Caching storage when read only once results in unnecessary PUSH and POP and leads to increased gas

Reply:
<img width="692" alt="Screenshot 2023-11-06 at 3 55 10 PM" src="https://github.com/bcnmy/biconomy-paymasters/assets/16562513/27367397-6381-40a4-b471-e3ed3e31384f">

>gas used: 27535


`+=` and `-=` seem to give a better result here. Depending on the bounds, you might want to use `unchecked` to save more gas if it's safe to do so.

Reply:
> <img alt="Screenshot 2023-11-06 at 3 47 21 PM" width="776" src="https://user-images.githubusercontent.com/16562513/280671206-faaa1c38-66d9-435c-85b2-3677f114be07.png">
> > gas used: 27646
> 
> <img alt="Screenshot 2023-11-06 at 3 49 01 PM" width="739" src="https://user-images.githubusercontent.com/16562513/280671719-0d7c9686-4983-44b3-80aa-8a72256062e8.png">
> > gas used: 27767
> 
> Caching storage when read only once results in unnecessary PUSH and POP and leads to increased gas

when I read this suggestion it my code it seemed like it's read twice.

Reply:
done